File: ios/Headstart/Core/HeadstartError.swift
// ios/Headstart/Core/HeadstartError.swift
import Foundation

// NOTE: the plan doc pulls in the FirebaseFunctions module here, purely to reach
// `FunctionsErrorDomain` and `FunctionsErrorCode`. Both are constants — the domain string
// and the gRPC status numbers — so they are declared below instead and this file stays
// Foundation-only, which keeps the mapper unit-testable with no Firebase linkage and
// satisfies the batch rule that nothing under Core/ depends on the Firebase SDK.

/// `FunctionsErrorDomain` from the Firebase Functions SDK.
public let kFunctionsErrorDomain = "com.firebase.functions"

/// The gRPC status codes `FunctionsErrorCode` is built from. Only the transport-level
/// facts that outrank the callable's message are needed here.
public enum CallableStatus {
    public static let deadlineExceeded = 4
    public static let unavailable = 14
    public static let unauthenticated = 16
}

/// Every failure the UI has to explain. `code` is the wire value from CLIENT_CONTRACT.md;
/// `userMessage` is what a screen may put in front of a person.
public enum HeadstartError: Error, Equatable, Sendable {
    case notPaired
    case tripActive
    case spotNotFound
    case badCode
    case ownCode
    case driverOnly
    case tripNotFound
    case badCoords
    case badName
    case badToken
    /// CLIENT_CONTRACT_ADDENDUM.md §O — `sendReply` with empty custom text.
    case badReply
    /// Spec §9 mentions rate limiting; not in the contract's exhaustive ten, but mapped.
    case rateLimited
    case unauthenticated
    case offline
    case unknown(String)

    public var code: String {
        switch self {
        case .notPaired: return "not-paired"
        case .tripActive: return "trip-active"
        case .spotNotFound: return "spot-not-found"
        case .badCode: return "bad-code"
        case .ownCode: return "own-code"
        case .driverOnly: return "driver-only"
        case .tripNotFound: return "trip-not-found"
        case .badCoords: return "bad-coords"
        case .badName: return "bad-name"
        case .badToken: return "bad-token"
        case .badReply: return "bad-reply"
        case .rateLimited: return "rate-limited"
        case .unauthenticated: return "unauthenticated"
        case .offline: return "offline"
        case .unknown(let raw): return raw
        }
    }

    public var userMessage: String {
        switch self {
        case .notPaired: return "You're not paired with anyone yet."
        case .tripActive: return "A trip is already running."
        case .spotNotFound: return "That pickup spot no longer exists."
        case .badCode: return "That code isn't valid. Check the six characters and try again."
        case .ownCode: return "That's your own invite code. Send it to the other person."
        case .driverOnly: return "Only the driver can do that."
        case .tripNotFound: return "That trip has already ended."
        case .badCoords: return "We couldn't read that location. Try again."
        case .badName: return "Give the spot a name of 1 to 40 characters."
        case .badToken: return "Notifications aren't set up on this phone yet."
        case .badReply: return "Write something before you send it."
        case .rateLimited: return "Too many tries. Wait a moment, then try again."
        case .unauthenticated: return "Sign in again to continue."
        case .offline: return "No connection. We'll try again when you're back online."
        case .unknown: return "Something went wrong. Try again."
        }
    }
}

/// Longest code first, so a `contains` match never returns a prefix of a longer code.
private let byCode: [HeadstartError] = [
    .notPaired, .tripActive, .spotNotFound, .badCode, .ownCode,
    .driverOnly, .tripNotFound, .badCoords, .badName, .badToken,
    .badReply, .rateLimited,
]

/// Pure mapper: callable message (plus two transport facts) to a typed error.
/// Auth and connectivity outrank the message, because a stale token can produce any message.
public func headstartErrorFor(
    _ message: String?,
    isNetwork: Bool = false,
    isUnauthenticated: Bool = false
) -> HeadstartError {
    if isUnauthenticated { return .unauthenticated }
    if isNetwork { return .offline }
    let text = (message ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
    if text.isEmpty { return .unknown("unknown") }
    if let exact = byCode.first(where: { $0.code == text }) { return exact }
    // Some transports prefix the status, e.g. "INVALID_ARGUMENT: bad-code".
    if let contained = byCode.first(where: { text.contains($0.code) }) { return contained }
    return .unknown(text)
}

public extension Error {
    /// Adapter from whatever the Firebase SDK threw.
    func asHeadstartError() -> HeadstartError {
        if let already = self as? HeadstartError { return already }
        let ns = self as NSError
        var isUnauthenticated = false
        var isNetwork = ns.domain == NSURLErrorDomain
        if ns.domain == kFunctionsErrorDomain {
            isUnauthenticated = ns.code == CallableStatus.unauthenticated
            isNetwork = isNetwork
                || ns.code == CallableStatus.unavailable
                || ns.code == CallableStatus.deadlineExceeded
        }
        let message = (ns.userInfo[NSLocalizedDescriptionKey] as? String) ?? ns.localizedDescription
        return headstartErrorFor(message, isNetwork: isNetwork, isUnauthenticated: isUnauthenticated)
    }
}

File: android/app/src/main/java/app/headstart/core/HeadstartError.kt
package app.headstart.core

import com.google.firebase.functions.FirebaseFunctionsException
import java.io.IOException

/**
 * Every failure the UI has to explain. [code] is the wire value from CLIENT_CONTRACT.md;
 * [userMessage] is what a screen may put in front of a person.
 */
sealed class HeadstartError(
    val code: String,
    val userMessage: String,
) : Exception(code) {
    object NotPaired : HeadstartError("not-paired", "You're not paired with anyone yet.")
    object TripActive : HeadstartError("trip-active", "A trip is already running.")
    object SpotNotFound : HeadstartError("spot-not-found", "That pickup spot no longer exists.")
    object BadCode : HeadstartError("bad-code", "That code isn't valid. Check the six characters and try again.")
    object OwnCode : HeadstartError("own-code", "That's your own invite code. Send it to the other person.")
    object DriverOnly : HeadstartError("driver-only", "Only the driver can do that.")
    object TripNotFound : HeadstartError("trip-not-found", "That trip has already ended.")
    object BadCoords : HeadstartError("bad-coords", "We couldn't read that location. Try again.")
    object BadName : HeadstartError("bad-name", "Give the spot a name of 1 to 40 characters.")
    object BadToken : HeadstartError("bad-token", "Notifications aren't set up on this phone yet.")

    /** Addendum §O: returned by `sendReply` when custom text is empty. */
    object BadReply : HeadstartError("bad-reply", "Type something before you send it.")

    object Unauthenticated : HeadstartError("unauthenticated", "Sign in again to continue.")
    object Offline : HeadstartError("offline", "No connection. We'll try again when you're back online.")

    // Client-side only: raised by AuthRepository, never returned by a callable.
    object InvalidPhone : HeadstartError("invalid-phone", "That doesn't look like a phone number we can text.")
    object InvalidSmsCode : HeadstartError("invalid-sms-code", "That code doesn't match. Check the six digits.")
    object SmsQuotaExceeded : HeadstartError("sms-quota", "Too many attempts. Wait a few minutes and try again.")
    object SessionExpired : HeadstartError("session-expired", "That code expired. Ask for a new one.")

    class Unknown(rawCode: String) : HeadstartError(rawCode, "Something went wrong. Try again.")
}

private val BY_CODE: List<HeadstartError> = listOf(
    HeadstartError.NotPaired,
    HeadstartError.TripActive,
    HeadstartError.SpotNotFound,
    HeadstartError.BadCode,
    HeadstartError.OwnCode,
    HeadstartError.DriverOnly,
    HeadstartError.TripNotFound,
    HeadstartError.BadCoords,
    HeadstartError.BadName,
    HeadstartError.BadToken,
    HeadstartError.BadReply,
)

/**
 * Pure mapper: callable message (plus two transport facts) to a typed error.
 * Auth and connectivity outrank the message, because a stale token can produce any message.
 */
fun headstartErrorFor(
    message: String?,
    isNetwork: Boolean = false,
    isUnauthenticated: Boolean = false,
): HeadstartError {
    if (isUnauthenticated) return HeadstartError.Unauthenticated
    if (isNetwork) return HeadstartError.Offline
    val text = message?.trim().orEmpty()
    if (text.isEmpty()) return HeadstartError.Unknown("unknown")
    BY_CODE.firstOrNull { it.code == text }?.let { return it }
    // Some transports prefix the status, e.g. "INVALID_ARGUMENT: bad-code".
    BY_CODE.firstOrNull { text.contains(it.code) }?.let { return it }
    return HeadstartError.Unknown(text)
}

/** Adapter from whatever the Firebase SDK threw. */
fun Throwable.toHeadstartError(): HeadstartError {
    if (this is HeadstartError) return this
    val fx = this as? FirebaseFunctionsException
    val isUnauthenticated = fx?.code == FirebaseFunctionsException.Code.UNAUTHENTICATED
    val isNetwork = this is IOException ||
        cause is IOException ||
        fx?.code == FirebaseFunctionsException.Code.UNAVAILABLE ||
        fx?.code == FirebaseFunctionsException.Code.DEADLINE_EXCEEDED
    return headstartErrorFor(message, isNetwork = isNetwork, isUnauthenticated = isUnauthenticated)
}
